pkg/controller/atlascustomrole/custom_role.go

```diff
@@ -0,0 +1,119 @@
+package atlascustomrole
+
+import (
+	"fmt"
+	"reflect"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/customroles"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+)
+
+type roleController struct {
+	ctx       *workflow.Context
+	projectID string
+	service   customroles.CustomRoleService
+	role      *akov2.AtlasCustomRole
+	dpEnabled bool
+	k8sClient client.Client
+}
+
+func handleCustomRole(ctx *workflow.Context, k8sClient client.Client, akoCustomRole *akov2.AtlasCustomRole,
+	dpEnabled bool) workflow.Result {
+	ctx.Log.Debug("starting custom role processing")
+	defer ctx.Log.Debug("finished custom role processing")
+
+	r := roleController{
+		ctx:       ctx,
+		service:   customroles.NewCustomRoles(ctx.SdkClient.CustomDatabaseRolesApi),
+		role:      akoCustomRole,
+		dpEnabled: dpEnabled,
+		k8sClient: k8sClient,
+	}
+
+	return r.Reconcile()
+}
+
+func (r *roleController) Reconcile() workflow.Result {
```

Comment:
this could be instead something like:

```go
func (r *AtlasCustomRoleReconciler) handle(customRoleRequest *CustomRoleRequest) {...
 ```
 
 And the request has anything that you need that is NOT already in the top level reconciler:
 ```go
type CustomRoleRequest struct {
	ctx       *workflow.Context
	projectID string
	role      *akov2.AtlasCustomRole
}
// service   customroles.CustomRoleService is already in the top level reconciler
// dpEnabled is already in the top level reconciler
// k8sClient client.Client is already in top level the reconciler

Which renders a request with just 3 elements, that could be just a call with 3 arguments:
[func (r *Resource) handle(ctx, projectID, Resource)](https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1930/files#diff-d17d85797ec5b7b965ec7b3508aab8f74d8bc04e930828939a234525efcaaddcR125)
```

Reply:
On the other hand, a request might make sense after all. The service(s), which might be more than one, cannot be solved until reconciliation starts, as they depend on credentials that depend on Kubernetes configuration that needs to be parse while doing the reconciliation.

So being strict, the services part is part of the request, and should not be set in the reconciler struct... which means we might have to fix this retroactively in several places.

Going back here, it seems the handle method should be invoked like this:
```go

r.handle(&ResourceHandleRequest{
  Context: ctx
  ProjectID: projectID:
  Resource: resource
  ServiceA: serviceA,...)
```

Note I am using generic naming as I think this can and should be generalized.

Reply:
CLOUDP-284792 about the services being a request context bound concept, not a reconciler field